pkg/ssl/sign.go

```diff
@@ -0,0 +1,260 @@
+/*
+Package ssl implements the Secure Systems Lab signing-spec (sometimes
+abbreviated SSL Siging spec.
+https://github.com/secure-systems-lab/signing-spec
+*/
+package ssl
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/binary"
+	"fmt"
+)
+
+// ErrUnknownKey indicates that the implementation does not recognize the
+// key.
+var ErrUnknownKey = fmt.Errorf("unknown key")
+
+// ErrNoSignature indicates that an envelope did not contain any signatures.
+var ErrNoSignature = fmt.Errorf("no signature found")
+
+/*
+Envelope captures an envelope as described by the Secure Systems Lab
+Signing Specification. See here:
+https://github.com/secure-systems-lab/signing-spec/blob/master/envelope.md
+*/
+type Envelope struct {
+	PayloadType string      `json:"payloadType"`
+	Payload     string      `json:"payload"`
+	Signatures  []Signature `json:"signatures"`
+}
+
+/*
+Signature represents a generic in-toto signature that contains the identifier
+of the key which was used to create the signature.
+The used signature scheme has to be agreed upon by the signer and verifer
+out of band.
+The signature is a base64 encoding of the raw bytes from the signature
+algorithm.
+*/
+type Signature struct {
+	KeyID string `json:"keyid"`
+	Sig   string `json:"sig"`
+}
+
+/*
+Pae implementes PASETO Pre-Authentic Encoding
+https://github.com/paragonie/paseto/blob/master/docs/01-Protocol-Versions/Common.md#authentication-padding
+*/
+func Pae(data [][]byte) ([]byte, error) {
+	var buf = bytes.Buffer{}
+	var l = len(data)
+	var err error
+
+	// Negative value sets the highest bit to 1.
+	// With two complements encoding, high bit must be set to 0 for
+	// interoperability with languages that lack unsigned integer types.
+	if !verifyPaeLength(int64(l)) {
+		return nil, fmt.Errorf("length must be less than 2^63")
+	}
+	if err = binary.Write(&buf, binary.LittleEndian, uint64(l)); err != nil {
+		return nil, err
+	}
+
+	for _, b := range data {
+		var bw int
+		l = len(b)
+
+		if !verifyPaeLength(int64(l)) {
```

Comment:
Do we need to do this check? I think the lengths all come from len() which probably guarantees it will be positive:

https://golang.org/pkg/builtin/#len

Reply:
Yeah, you are correct; from the spec: `0 <= len(s) <= cap(s)`. I'll remove the checks.